### PR TITLE
[Datasets] Last-mile preprocessing docs.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Datasets docs for last-mile preprocessing, particularly geared towards ML ingest. This gives groupby, aggregations, and random shuffling examples in the overview page (not present previously), adds some concreteness to our last-mile preprocessing positioning, and provides some preprocessing recipes for a few common transformations.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/dataset.rst [hunk 1]
```diff
@@ -327,6 +327,134 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+    from ray.data.aggregate import Mean, Std
+    ds.aggregate(Mean("B"), Std("B", ddof=0), Mean("C"), Std("C", ddof=0))
+    # -> {'mean(A)': 0.9, 'std(A)': 0.8306623862918076, 'mean(B)': 9.0, 'std(B)': 5.744562646538029}
+
+These aggregations can be combined with batch mapping to e.g. efficiently impute missing values with statistics and standard scale feature columns.
+
+.. code-block:: python
+
+    # Impute missing values with the column mean.
+    b_mean = ds.mean("B")
```
Comment (clarkzinzow) [inline doc/source/data/dataset.rst:405]: Note that this will currently fail if nulls exist in column `"B"`, since our aggregations aren't resilient to null values. I'm going to submit a sibling PR that makes our statistical aggregation resilient to nulls by default.

File: doc/source/data/dataset.rst [hunk 2]
```diff
@@ -327,6 +327,134 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+    ds = ray.data.range(10)
+    # -> [0, 1, ..., 9]
+
+    # Global random shuffle.
+    ds = ds.random_shuffle()
+    # -> [6, 2, ..., 4]
+
+    # Scales to terabytes of data with the same simple API.
+    ds = ray.data.read_parquet("s3://ursa-labs-taxi-data")  # open, tabular, NYC taxi dataset
+    # -> Dataset(num_blocks=2000, num_rows=4000000000, schema={col1: int64, col2: string})
```
Comment (clarkzinzow) [inline doc/source/data/dataset.rst:445]: These aren't the actual number of blocks/rows for this dataset, should probably align these.

File: doc/source/data/dataset.rst [hunk 3]
```diff
@@ -327,6 +327,134 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+
+    ds = ray.data.range(10)
+    # -> [0, 1, ..., 9]
+
+    # Global random shuffle.
+    ds = ds.random_shuffle()
+    # -> [6, 2, ..., 4]
+
+    # Scales to terabytes of data with the same simple API.
+    ds = ray.data.read_parquet("s3://ursa-labs-taxi-data")  # open, tabular, NYC taxi dataset
```
Comment (clarkzinzow) [inline doc/source/data/dataset.rst:444]: Decided that a real, open dataset would be good in case users are wanting to quickly try it out.

File: doc/source/data/dataset.rst [hunk 4]
```diff
@@ -327,6 +327,134 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+        df["new_col"] = df["feature_1"] - 2 * df["feature_2"] + df["feature_3"] / 3
+        # Transform existing column.
+        df["feature_1"] = 2 * df["feature_1"] + 1
+        # Drop column.
+        df.drop(columns="feature_2", inplace=True)
+        # One-hot encoding.
+        fruits = ["apple", "orange", "banana"]
+        for fruit in fruits:
+            df[f"fruit_{fruit}"] = df[fruit].map(
+                collections.defaultdict(int, **{fruit: 1}))
```
Comment (clarkzinzow) [inline doc/source/data/dataset.rst:362]: The columns in this transformation are disjoint from the columns in the groupby + aggregations and random shuffling sections, could unify these to a common data model if that would be cleaner for the user to digest.

Comment (worldveil) [reply doc/source/data/dataset.rst:363]: could we make the column something like "state_<state abbr>" or something? might be clearer to user. fruit just comes out of nowhere

File: doc/source/data/dataset.rst [hunk 5]
```diff
@@ -327,6 +327,135 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+        return df
+
+    ds = ds.map_batches(transform_batch, batch_format="pandas")
```
Comment (worldveil) [inline doc/source/data/dataset.rst:366]: comment about what "batch_format" is here would be helpful.

File: doc/source/data/dataset.rst [hunk 6]
```diff
@@ -327,6 +327,135 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+
+Groupbys and aggregations
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Other preprocessing operations require global operations, such as groupbys and grouped/global aggregations.
+
+.. code-block:: python
+
+    ds = ray.data.from_items([
```
Comment (worldveil) [inline doc/source/data/dataset.rst:375]: we should clarify that the type here is Ray Datasets, not a pandas DF

File: doc/source/data/dataset.rst [hunk 7]
```diff
@@ -327,6 +327,135 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+        {"A": x % 3, "B": 2 * x, "C": 3 * x}
+        for x in range(10)])
+
+    # Group by the A column and calculate the per-group mean for B and C columns.
+    agg_ds = ds.groupby("A").mean(["B", "C"])
+    agg_ds.to_pandas()
```
Comment (worldveil) [inline doc/source/data/dataset.rst:381]: is everything lazy up until this line?

File: doc/source/data/dataset.rst [hunk 8]
```diff
@@ -327,6 +327,135 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+    b_mean = ds.mean("B")
+
+    def impute_b(df: pd.DataFrame):
+        df["B"].fillna(b_mean)
+        return df
+
+    ds = ds.map_batches(impute_b, batch_format="pandas")
+
+    # Standard scaling of all feature columns.
+    stats = ds.aggregate(Mean("B"), Std("B"), Mean("C"), Std("C"))
```
Comment (worldveil) [inline doc/source/data/dataset.rst:415]: where are these functions coming from ? `Mean`, `Std` ?

File: doc/source/data/dataset.rst [hunk 9]
```diff
@@ -327,6 +327,135 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+    def batch_standard_scaler(df: pd.DataFrame):
+        def column_standard_scaler(s: pd.Series):
+            if s.name == "label":
+                # Don't standard scale label column.
+                return s
+            s_mean = stats[f"mean({s.name})"]
+            s_std = stats[f"std({s.name})"]
+            return (s - s_mean) / s_std
+
+        return df.transform(column_standard_scaler)
```
Comment (worldveil) [inline doc/source/data/dataset.rst:426]: does this operate on every column of the dataframe?

File: doc/source/data/dataset.rst [hunk 10]
```diff
@@ -327,6 +327,135 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+
+    ds = ds.map_batches(batch_standard_scaler, batch_format="pandas")
+
+Random shuffle
+~~~~~~~~~~~~~~
+
+Randomly shuffling data is an important part of training machine learning models: it decorrelates samples, preventing overfitting and improving generalization. For many models, even between-epoch shuffling can drastically improve the precision gain per step/epoch. Datasets has a hyper-scalable distributed random shuffle:
+
+.. code-block:: python
```
Comment (worldveil) [inline doc/source/data/dataset.rst:435]: add imports here?

File: doc/source/data/dataset.rst [hunk 11]
```diff
@@ -327,6 +327,135 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+    # -> Dataset(num_blocks=2000, num_rows=4000000000, schema={col1: int64, col2: string})
+    ds = ds.random_shuffle()
+    # -> Dataset(num_blocks=2000, num_rows=4000000000, schema={col1: int64, col2: string})
+
+    # Per-epoch shuffling is as simple as changing where we shuffle.
+    num_epochs = 20
+    # Shuffle once.
+    ds.random_shuffle().repeat(num_epochs)
+    # Shuffle every epoch.
+    ds.repeat(num_epochs).random_shuffle()
```
Comment (worldveil) [inline doc/source/data/dataset.rst:455]: `ds.random_shuffle().repeat(num_epochs)` vs `ds.repeat(num_epochs).random_shuffle()` is a weirdly subtle difference. can we explain this somehow?

File: doc/source/data/dataset.rst [hunk 12]
```diff
@@ -327,6 +327,135 @@ By default, transformations are executed using Ray tasks. For transformations th
...
+    # Global mean on multiple columns.
+    ds.mean(["B", "C"])
+    # -> {'mean(B)': 9.0, 'mean(C)': 13.5} 
+
+    # Multiple global aggregations on multiple columns.
+    from ray.data.aggregate import Mean, Std
+    ds.aggregate(Mean("B"), Std("B", ddof=0), Mean("C"), Std("C", ddof=0))
+    # -> {'mean(A)': 0.9, 'std(A)': 0.8306623862918076, 'mean(B)': 9.0, 'std(B)': 5.744562646538029}
+
+These aggregations can be combined with batch mapping to e.g. efficiently impute missing values with statistics and standard scale feature columns.
```
Comment (scv119) [inline doc/source/data/dataset.rst:401]: sentence looks broken?

File: doc/source/data/dataset.rst (hunk 4 above)
Comment (clarkzinzow) [reply doc/source/data/dataset.rst:363]: I think that I'll change this to `categories = ["cat_1", "cat_2", "cat_3"]` and `df[f"categorical_{cat}"]`, to make it extra clear that this is a categorical column.

File: doc/source/data/dataset.rst (hunk 7 above)
Comment (clarkzinzow) [reply doc/source/data/dataset.rst:381]: Everything is eager, actually; we highlight the eagerness of transformations in the "Transforming Datasets" section, but I suppose that we should explicitly call out that aggregations are eager. I'll add that to this section!

File: doc/source/data/dataset.rst (hunk 12 above)
Comment (clarkzinzow) [reply doc/source/data/dataset.rst:401]: How so?

File: doc/source/data/dataset.rst (hunk 8 above)
Comment (clarkzinzow) [reply doc/source/data/dataset.rst:415]: They are imported (and used) in the previous example.

File: doc/source/data/dataset.rst (hunk 9 above)
Comment (clarkzinzow) [reply doc/source/data/dataset.rst:426]: Yes, which is why `column_standard_scaler` manually checks for the `"label"` column to make sure that we don't scale it. An alternative column selection method would be something like this:
```python
    def batch_standard_scaler(df: pd.DataFrame):
        def column_standard_scaler(s: pd.Series):
            s_mean = stats[f"mean({s.name})"]
            s_std = stats[f"std({s.name})"]
            return (s - s_mean) / s_std

        cols = df.columns.difference(["label"])
        return df.loc[cols].transform(column_standard_scaler)
```

I think that this is more readable, but will possibly result in an extra copy of the DataFrame with `df.loc[cols]`. I'll make the change sinceI think the former outweighs the latter.

File: doc/source/data/dataset.rst (hunk 10 above)
Comment (clarkzinzow) [reply doc/source/data/dataset.rst:435]: This overview/quick start page currently doesn't duplicate imports across example snippets in an effort to reduce redundancy and make the examples more readable, and the only needing import for this snippet is `import ray`. I think that we should stick with this import scheme; this seems to be standard practice for similar libraries: [NumPy](https://numpy.org/doc/1.21/user/absolute_beginners.html), [Pandas](https://pandas.pydata.org/docs/user_guide/10min.html), [PySpark](https://spark.apache.org/docs/latest/sql-getting-started.html).

File: doc/source/data/dataset.rst (hunk 11 above)
Comment (clarkzinzow) [reply doc/source/data/dataset.rst:455]: I'm attempting to explain it here with this example, can you explain what's not clear?

Here is the more verbose version that I originally had for this example:
```python
    # Per-epoch shuffling is as simple as changing where we invoke the shuffle:
    #   - Before repeating => dataset is shuffled once.
    #   - After repeating    => dataset is shuffled before ever epoch.
    num_epochs = 20
    # Shuffle once, before the dataset is repeated for num_epochs epochs.
    ds.random_shuffle().repeat(num_epochs)
    # Shuffle repeatedly, before every epoch.
    ds.repeat(num_epochs).random_shuffle_each_window()
```

Is that more clear?

File: doc/source/data/dataset.rst [hunk 13]
```diff
@@ -327,6 +327,135 @@ By default, transformations are executed using Ray tasks. For transformations th
     # Save the results.
     ds.repartition(1).write_json("s3://bucket/inference-results")
 
+Last-mile preprocessing
```
Comment (ericl) [inline doc/source/data/dataset.rst:330]: How about we move this into a "Dataset ML preprocessing" page before/after the tensor support one? (the overview page is getting too large, and also should be split up later)

File: doc/source/data/dataset.rst (hunk 12 above)
Comment (scv119) [reply doc/source/data/dataset.rst:401]: the e.g. looks a bit confusing to me.

File: doc/source/data/dataset.rst (hunk 13 above)
Comment (clarkzinzow) [reply doc/source/data/dataset.rst:330]: @ericl Can we do that in a future PR? We can do that as a docs polish PR during the GA push. A big motivation for this PR is to get groupby + aggregations representation in the 1.9 docs, particularly in our overview/getting started/user guide page, and I think reorganizing these sections into separate pages is out of scope for that.

**Comment by ericl** [comment]
I'd like to start putting these advanced ops in separate pages. You can
link to it from the overview.

**Comment by clarkzinzow** [comment]
@ericl Getting groupby + aggregations into our existing overview docs page before the 1.9 release is a P0 verging on a release blocker, while reorganizing the overview docs page is a P1 in my mind and I don't want that to block this from getting merged. Do you disagree with that prioritization?

**Comment by ericl** [comment]
@clarkzinzow , I'm suggesting moving the new section to a new page, not re-organizing the docs. I think this will be a strict improvement over the current PR, which makes the overview too long / detailed.

We can revisit a more holistic refactoring separately.

**Comment by clarkzinzow** [comment]
@ericl Done, please review.

**Review by ericl — APPROVED** [review]
Looks good, but could we call it ML preprocessing instead of last mile? I think that will be the more common search term.

**Comment by clarkzinzow** [comment]
>Looks good, but could we call it ML preprocessing instead of last mile? I think that will be the more common search term.

Sure, no concerns that removing last-mile will confuse our positioning?

**Comment by iycheng** [comment]
Merge it since the doc has been built.
